build: Add CMake-based build system (2 of N)

hebasto opened the pull request:
The parent PR: https://github.com/bitcoin/bitcoin/pull/25797.
The previous PR in the staging branch: https://github.com/hebasto/bitcoin/pull/5.

This PR details:

1. Added project-wide system introspection.

| Autoconf macro | Symbols in `bitcoin-config.h` | CMake implementation |
|---|---|---|
| [`AC_INIT`](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.69/autoconf.html#Initializing-configure) | `PACKAGE_NAME` `PACKAGE_TARNAME` `PACKAGE_VERSION` `PACKAGE_STRING` `PACKAGE_BUGREPORT` `PACKAGE_URL` | Done in https://github.com/hebasto/bitcoin/pull/5, except for `PACKAGE_TARNAME` and `PACKAGE_STRING`, which are not currently used |
| [`AX_CXX_COMPILE_STDCXX`](https://www.gnu.org/software/autoconf-archive/ax_cxx_compile_stdcxx.html) | `HAVE_CXX17` `HAVE_CXX20` | CMake uses its own [implementation](https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html) |
| [`AC_C_BIGENDIAN`](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.69/autoconf.html#index-AC_005fC_005fBIGENDIAN-892) | `AC_APPLE_UNIVERSAL_BUILD` `WORDS_BIGENDIAN` and C headers | CMake uses its own [implementation](https://cmake.org/cmake/help/latest/module/TestBigEndian.html) |
| [`AX_PTHREAD`](https://www.gnu.org/software/autoconf-archive/ax_pthread.html) | `HAVE_PTHREAD` `HAVE_PTHREAD_PRIO_INHERIT` `PTHREAD_CREATE_JOINABLE` | CMake uses its own [implementation](https://cmake.org/cmake/help/latest/module/FindThreads.html) |
| [`AC_SYS_LARGEFILE`](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.69/autoconf.html#index-AC_005fSYS_005fLARGEFILE-1103) | `_FILE_OFFSET_BITS` `_LARGE_FILES` | _Not implemented yet_ |
| [`AC_FUNC_STRERROR_R`](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.69/autoconf.html#index-AC_005fFUNC_005fSTRERROR_005fR-517) | `HAVE_DECL_STRERROR_R` `HAVE_STRERROR_R` `STRERROR_R_CHAR_P` and C headers  | Implemented `STRERROR_R_CHAR_P`, others are not used
| C headers |  `HAVE_INTTYPES_H` `HAVE_STDINT_H` `HAVE_STDIO_H` `HAVE_STDLIB_H` `HAVE_STRINGS_H` `HAVE_STRING_H` `HAVE_SYS_STAT_H` `HAVE_SYS_TYPES_H` `HAVE_UNISTD_H` `STDC_HEADERS` | Only `HAVE_STRING_H`, `HAVE_SYS_TYPES_H` and `HAVE_UNISTD_H` are evaluated for internal needs

---

2. Added the `leveldb` build target (including its `crc32c` dependency):

- on Linux / *BSD / macOS:
```sh
cmake -S . -B build
cmake --build build --target leveldb
```

- on Windows:
```cmd
cmake -G "Visual Studio 17 2022" -A x64 -S . -B build
cmake --build build --config Debug --target leveldb  # or --config Release
```

**NOTE**. On Windows (MSVC), the `leveldb` library now uses `crc32c` one. That is not the case when using the current build system from the `build_msvc` directory: https://github.com/hebasto/bitcoin/blob/fe1b3256888bd0e70d0c9655f565e139ec87b606/build_msvc/libleveldb/libleveldb.vcxproj#L53

hebasto commented:
The PR description has been updated with details about CMake's implementation of Autoconf's system introspection macros.

@theuni [wrote](https://github.com/hebasto/bitcoin/pull/5#issuecomment-1431851225):
> and have the _next_ PR be the big one that brings in the structure and at least one build target.

This PR delivers a few build targets. But I couldn't move to actual building without tiding up system introspection. Sorry if it didn't meet your expectations. But I promise the next PR will deliver `bitcoind` target.

hebasto commented:
Friendly ping (as GitHub, probably, won't) people who [showed](https://github.com/bitcoin/bitcoin/pull/25797) some interest in CMake:

@achow101 @adam2k @dongcarl @fanquake @jarolrod @luke-jr @MarcoFalke @ryanofsky @sipsorcery @TheCharlatan @theStack @theuni @vasild @willcl-ark

theuni commented:
At a high-level, I think we need some temporary place to stash autotools parity checks like `introspection_compat.cmake`.

Reason being: autoconf includes lots of ancient checks that aren't worth the time and effort anymore. Good example is `HAVE_STRING_H`.

So imo, it'd be helpful for you to be able to make the distinction of "I'm adding this to keep parity with autotools and keep the cmake/bitcoin-config.h identical, but it's not longer necessary and we can remove it all at some point after CMake is merged"

fanquake commented:
> autoconf includes lots of ancient checks that aren't worth the time and effort anymore. Good example is HAVE_STRING_H.

Yep. I see little reason to port all of this cruft over into CMake (or keep it for the long term), especially when we don't even do anything based on the output of most tests.

theuni commented:
> Yep. I see little reason to port all of this cruft over into CMake (or keep it for the long term), especially when we don't even do anything based on the output of most tests.

I suspect @hebasto's motivation here is trying to exactly re-create `bitcoin-config.h` exactly as before. IMO that's quite reasonable, as without that we'd never be able to compare object output before/after.

But yes, I very much agree that keeping it for the long term is unnecessary and/or harmful.

hebasto commented:
@theuni
> At a high-level, I think we need some temporary place to stash autotools parity checks like `introspection_compat.cmake`.
> 
> Reason being: autoconf includes lots of ancient checks that aren't worth the time and effort anymore. Good example is `HAVE_STRING_H`.
> 
> So imo, it'd be helpful for you to be able to make the distinction of "I'm adding this to keep parity with autotools and keep the cmake/bitcoin-config.h identical, but it's not longer necessary and we can remove it all at some point after CMake is merged"

Agree. For now, only required / used in our code definitions are evaluated and passed into the `bitcoin-config.h` header.

`HAVE_STRING_H` is used just internally for evaluating of `STRERROR_R_CHAR_P`: https://github.com/hebasto/bitcoin/blob/eeb8465444af566d98dced950fbd48cb7653fbfa/cmake/introspection.cmake#L134-L147

The same way, internally only, some other definitions are used.

hebasto commented:
> I suspect @hebasto's motivation here is trying to exactly re-create `bitcoin-config.h` exactly as before.

Exactly, as it was explicitly [requested](https://github.com/bitcoin/bitcoin/pull/27060#discussion_r1102774935).

But I'm not going to pull unused crap. That is why a table in the PR description explains which definition is absent and why.

theuni commented:
> > I suspect @hebasto's motivation here is trying to exactly re-create `bitcoin-config.h` exactly as before.
> 
> Exactly, as it was explicitly [requested](https://github.com/bitcoin/bitcoin/pull/27060#discussion_r1102774935).
> 
> But I'm not going to pull unused crap. That is why a table in the PR description explains which definition is absent and why.

Yes, this is very helpful and I appreciate the approach. A diff between the old/new `bitcoin-config.h` looks as expected for me given the chart above.

theuni commented on cmake/introspection.cmake:
```diff
@@ -17,3 +19,215 @@ CHECK_INCLUDE_FILE_CXX(sys/select.h HAVE_SYS_SELECT_H)
+
+# Check for gmtime_r(), fallback to gmtime_s() if that is unavailable.
+# Fail if neither are available.
+check_cxx_source_compiles("
```
These are actually behavioral changes because CMake (surprisingly imo) invokes the linker for `check_cxx_source_compiles`.

We may want to investigate `set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)` at the top-level to disable that, then opting-in to linking as necessary (like for the socket check above).

hebasto replied:
[Updated](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1436930235).

theuni commented on cmake/introspection.cmake:
```diff
@@ -17,3 +19,215 @@ CHECK_INCLUDE_FILE_CXX(sys/select.h HAVE_SYS_SELECT_H)
+  endif()
+endif()
+
+if(WIN32)
```
I'd rather not start special-casing things like this. They tend to get outdated very quickly that way. All we need to know here is "`is std::system` or `::_wsystem` available?". We don't need to worry about which platforms to tie those checks to.

Doing it that way would continue to work in a future where, for example, `wsystem` is removed in windows and replaced with `std::system`. In that case, the check here would still pass as it should.

hebasto replied:
Thanks! [Updated](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1435777703).

theuni commented on cmake/introspection.cmake:
```diff
@@ -17,3 +19,215 @@ CHECK_INCLUDE_FILE_CXX(sys/select.h HAVE_SYS_SELECT_H)
+" HAVE_SYS_GETRANDOM)
+
+# - BSD getentropy()
+check_cxx_symbol_exists(getentropy "sys/random.h" HAVE_GETENTROPY_RAND)
```
Autotools also has `#include <unistd.h>`. Missing here?

hebasto replied:
Thanks! [Fixed](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1435777703).

theuni commented on cmake/introspection.cmake:
```diff
@@ -231,3 +231,25 @@ int main()
+int main(){}
+" HAVE_DEFAULT_VISIBILITY_ATTRIBUTE)
+
+if(WIN32)
```
Again, no need to special-case this to Win32. One day `__declspec(dllexport)` may work for FreeBSD, who knows? :)

hebasto replied:
[Updated](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1435777703).

vasild replied:
> One day `__declspec(dllexport)` may work for FreeBSD

:open_mouth: :angry: :fire: :ice_cube: :no_entry: ~!*@&^(_ !!!

theuni commented on src/CMakeLists.txt:
```diff
@@ -18,6 +18,54 @@ mark_as_advanced(
+set(HAVE_CRC32C ON CACHE BOOL "" FORCE)
+set(HAVE_SNAPPY OFF CACHE BOOL "" FORCE)
+set(HAVE_TCMALLOC OFF CACHE BOOL "" FORCE)
+# Disabling TSA warnings because leveldb simultaneously
```
I don't understand this comment? Is there a TSA warning that turns into an error?

hebasto replied:
This code is [dropped](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1435777703) now.

theuni commented on src/CMakeLists.txt:
```diff
@@ -18,6 +18,54 @@ mark_as_advanced(
+)
+target_compile_options(leveldb
+  PRIVATE
+    # RTTI has been forcibly disabled in leveldb's CMake configuration
```
Exceptions are also disabled there, I assume that would cause sneaky problems as well?

This seems brittle to me.

hebasto replied:
This code is [dropped](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1435777703) now.

theuni commented:
Another high-level comment: it might be easier to copy our existing way of building crc32/leveldb (basically reinventing their buildsystem internally) than to battle all of the new changes that using the upstream CMake buildsystem drags in.

In fact, yeah, thinking about it some more I'd much prefer it that way for our initial CMake merge. Then we could switch to the upstream buildsystems at a later date and with more review scrutiny.

hebasto commented on cmake/introspection.cmake:
```diff
@@ -17,3 +19,215 @@ CHECK_INCLUDE_FILE_CXX(sys/select.h HAVE_SYS_SELECT_H)
+    check_cxx_source_compiles("${check_socket_source}" IFADDR_WORKS_WITHOUT_LINK)
+    if(NOT IFADDR_WORKS_WITHOUT_LINK)
+      set(CMAKE_REQUIRED_LIBRARIES socket)
+      check_cxx_source_compiles("${check_socket_source}" IFADDR_NEEDS_LINK_TO_SOCKET)
```
> Why not use `CMAKE_REQUIRED_LIBRARIES` in check_cxx_source_compiles instead of setting/unsetting it?

Mind elaborating, please?

theuni commented:
> Another high-level comment: it might be easier to copy our existing way of building crc32/leveldb (basically reinventing their buildsystem internally) than to battle all of the new changes that using the upstream CMake buildsystem drags in.

I'm currently working on converting leveldb to be like this to see how it looks, I'll push it up when done.

hebasto commented:
> > Another high-level comment: it might be easier to copy our existing way of building crc32/leveldb (basically reinventing their buildsystem internally) than to battle all of the new changes that using the upstream CMake buildsystem drags in.
> 
> I'm currently working on converting leveldb to be like this to see how it looks, I'll push it up when done.

Maybe my old [branch](https://github.com/hebasto/bitcoin/tree/cmake/0615.4/cmake) can be helpful:
- [`cmake/subtree-crc32c.cmake`](https://github.com/hebasto/bitcoin/blob/cmake/0615.4/cmake/subtree-crc32c.cmake)
- [`cmake/subtree-leveldb.cmake`](https://github.com/hebasto/bitcoin/blob/cmake/0615.4/cmake/subtree-leveldb.cmake)

?

hebasto commented:
@furszy @0xB10C @w0xlt 

Mind joining us here for testing and reviewing?

theuni commented:
> Maybe my old [branch](https://github.com/hebasto/bitcoin/tree/cmake/0615.4/cmake) can be helpful:

Hah, thanks but too late. Here's my attempt:
https://github.com/theuni/bitcoin/commits/cmake2-selfleveldb

I like this approach MUCH better for now. I think we should work on switching to the upstream buildsystem eventually, but this is *such* a simpler change to deal with.

hebasto commented:
@theuni 
> > Maybe my old [branch](https://github.com/hebasto/bitcoin/tree/cmake/0615.4/cmake) can be helpful:
> 
> Hah, thanks but too late. Here's my attempt: https://github.com/theuni/bitcoin/commits/cmake2-selfleveldb
> 
> I like this approach MUCH better for now. I think we should work on switching to the upstream buildsystem eventually, but this is _such_ a simpler change to deal with.

Thank you very much for your suggestion! It has been implemented now with a few differences as follows:
- a structure of a target follows the pattern used in the parent https://github.com/bitcoin/bitcoin/pull/25797. i.e, source files, with no headers - definitions - include directories - compile options - link libraries - link options
- added a missed `src/leveldb/util/histogram.cc` source file
- [generator expressions](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html) are used for brevity
- adjusted for MSVC
- `include()` command has been moved into the top `CMakeLists.txt` to avoid unneeded contamination by project-wide definitions and options which are set in the `src/CMakeLists.txt`

---

Other @theuni's comments have been addressed as well.

The PR description has been updated with MSVC-related details.

hebasto commented on cmake/introspection.cmake:
```diff
@@ -17,3 +19,215 @@ CHECK_INCLUDE_FILE_CXX(sys/select.h HAVE_SYS_SELECT_H)
+        freeifaddrs(ifaddr);
+      }
+    ")
+    check_cxx_source_compiles("${check_socket_source}" IFADDR_WORKS_WITHOUT_LINK)
```
Thanks! [Updated](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1435777703).

hebasto commented:
**A note for reviewers.**

Consider compiling:
- with CMake:
```sh
cmake -S . -B build -DCMAKE_CXX_FLAGS="-g -O2"
cmake --build build --target leveldb
```
and
- with Autotools:
```sh
./autogen.sh
./configure --disable-hardening
make clean
make -C src leveldb/libleveldb.la leveldb/libmemenv.la
```

The resulted CMake's `build/libleveldb.a` is essentially the same as the combined Autotools' `src/leveldb/.libs/libleveldb.a` and `src/leveldb/.libs/libmemenv.a`.

fanquake commented on cmake/introspection.cmake:
```diff
@@ -0,0 +1,255 @@
+" HAVE_SYSCTL_ARND)
+
+check_cxx_source_compiles("
+#include <stdint.h>
```
Any reason we can't use the c++ headers here?

hebasto replied:
Thanks! [Fixed](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1436835797).

fanquake commented on cmake/introspection.cmake:
```diff
@@ -0,0 +1,255 @@
+CHECK_INCLUDE_FILE_CXX(sys/prctl.h HAVE_SYS_PRCTL_H)
+CHECK_INCLUDE_FILE_CXX(sys/resources.h HAVE_SYS_RESOURCES_H)
+CHECK_INCLUDE_FILE_CXX(sys/select.h HAVE_SYS_SELECT_H)
+CHECK_INCLUDE_FILE_CXX(sys/sysctl.h HAVE_SYS_SYSCTL_H)
```
We check for `sys/sysctl.h` here, and set `HAVE_SYS_SYSCTL_H`, but then never use `HAVE_SYS_SYSCTL_H`, while including `sys/sysctl.h` unconditionally inside CMake, and under various different identifiers in our src, i.e `HAVE_SYSCTL`, `HAVE_SYSCTL_ARND`, `MAC_OSX && ENABLE_ARM_SHANI`.

Similar for `sys/select.h`. We do a header check, but then never use `HAVE_SYS_SELECT_H` and just include `sys/select.h` in a `#ifndef WIN32` in compat.

I think now would be the right time to introduce some consistency. i.e we could drop the header checks here, and assume that if we can use `sysctl()`/`sysctlbyname()`, then we can include `sys/sysctl.h`, and use the same define for both, or, we could actually use the output of the header checks to include the headers. I don't think we should port checks that create unused output, and then go and do a number of different things in the source anyways. 

Doing this for a "similar bitcoin-config.h" is fine (either initially/during first review/porting), but no point bringing this all in at merge-time if it's unused, or not without a plan for how we are going to consolidate/remove it all.

hebasto replied:
Thanks! [Updated](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1436835797).

fanquake commented on cmake/leveldb.cmake:
```diff
@@ -0,0 +1,92 @@
+)
+
+#TODO: figure out how to filter out:
+# -Wconditional-uninitialized -Werror=conditional-uninitialized -Wsuggest-override -Werror=suggest-override
```
Can we atleast just turn off -Werror, there's no need to have that enabled.

hebasto replied:
We can return to this comment when introducing default warnings and `-DWERROR` option.

theuni replied:
Agreed. This is ugly for now, but I'm ok punting on this until we see how warning/error handling is intended to be... handled :)

fanquake commented on src/CMakeLists.txt:
```diff
@@ -2,6 +2,22 @@
+set(CRC32C_INSTALL OFF CACHE BOOL "" FORCE)
+# Honor visibility properties for all target types.
+# Required for Crc32c sub-project as it has the CMake minimum
+# required version 3.1.
```
How much cleaner is this going to be if we just bump the CMake minimum required version in the [crc32c project](https://github.com/bitcoin-core/crc32c-subtree)?

hebasto replied:
> How much cleaner is this going to be if we just bump the CMake minimum required version in the [crc32c project](https://github.com/bitcoin-core/crc32c-subtree)?

Just one LOC `set(CMAKE_POLICY_DEFAULT_CMP0063 NEW)` will gone.

Much cleaner it would be if defaults of `CRC32C_*` options are set to `OFF` in the [crc32c project](https://github.com/bitcoin-core/crc32c-subtree).

hebasto replied:
This code has been [dropped](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1436835797).

theuni commented:
I'd strongly prefer to integrate crc32c into our buildsystem like this instead for now: https://github.com/hebasto/bitcoin/blob/cmake/0615.4/cmake/subtree-crc32c.cmake (unreviewed in detail, but concept ACK :)

Again, I'm NOT advocating for keeping it that way long-term, I think your approach here is the way to go eventually. But for now, it's effectively forcing reviewers to review the entire crc32c buildsystem as well, leading to questions like: https://github.com/hebasto/bitcoin/pull/6#discussion_r1111255501

hebasto commented:
@theuni 
> I'd strongly prefer to integrate crc32c into our buildsystem like this instead for now: https://github.com/hebasto/bitcoin/blob/cmake/0615.4/cmake/subtree-crc32c.cmake (unreviewed in detail, but concept ACK :)
> 
> Again, I'm NOT advocating for keeping it that way long-term, I think your approach here is the way to go eventually. But for now, it's effectively forcing reviewers to review the entire crc32c buildsystem as well, leading to questions like: [#6 (comment)](https://github.com/hebasto/bitcoin/pull/6#discussion_r1111255501)

Done.

Also addressed @fanquake's comments:
- https://github.com/hebasto/bitcoin/pull/6#discussion_r1111247431
- https://github.com/hebasto/bitcoin/pull/6#discussion_r1111253359

hebasto commented:
@theuni 
> These are actually behavioral changes because CMake (surprisingly imo) invokes the linker for `check_cxx_source_compiles`.
> 
> We may want to investigate `set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)` at the top-level to disable that, then opting-in to linking as necessary (like for the socket check above).

Thanks! A dedicated `check_cxx_source_links` macro has been added.

hebasto commented:
> @theuni [wrote](https://github.com/hebasto/bitcoin/pull/5#issuecomment-1431851225):
> 
> > and have the _next_ PR be the big one that brings in the structure and at least one build target.
> 
> ... I promise the next PR will deliver `bitcoind` target.

See https://github.com/hebasto/bitcoin/pull/7 :)

TheCharlatan commented on cmake/bitcoin-config.h.in:
```diff
@@ -63,6 +180,9 @@
 
+/* Define to 1 if strerror_r returns char *. */
+#cmakedefine STRERROR_R_CHAR_P 1
+
```
Going through all the symbols defined in here and could not find the following builtins from our current bitcoin-config.h.in:
`HAVE_BUILTIN_CLZL` `HAVE_BUILTIN_CLZLL` `HAVE_BUILTIN_MUL_OVERFLOW`.

The crypto intrinsics are accounted for in the next PR, but I couldn't find `HAVE_TIMINGSAFE_BCMP`.

I also could not find `HAVE_EVHTTP_CONNECTION_GET_PEER_CONST_CHAR` and `HAVE_THREAD_LOCAL`.

My guess is I'm missing something here. The other symbols not carried over from the old config are indeed unused.

hebasto replied:
@TheCharlatan 

Thank you so much!

> Going through all the symbols defined in here and could not find the following builtins from our current bitcoin-config.h.in: `HAVE_BUILTIN_CLZL` `HAVE_BUILTIN_CLZLL` `HAVE_BUILTIN_MUL_OVERFLOW`.

Fixed.

> The crypto intrinsics are accounted for in the next PR, but I couldn't find `HAVE_TIMINGSAFE_BCMP`.

Going to add it to next PR.

> I also could not find `HAVE_EVHTTP_CONNECTION_GET_PEER_CONST_CHAR` and `HAVE_THREAD_LOCAL`.

Same. 

`HAVE_EVHTTP_CONNECTION_GET_PEER_CONST_CHAR` should be an amend of `pkg_check_modules(libevent)`, and `HAVE_THREAD_LOCAL` should be added to `find_package(Threads)`.

theuni replied:
For `HAVE_THREAD_LOCAL` see `target_compile_features( ... cxx_thread_local ... )`.

TheCharlatan replied:
The [diff](https://github.com/hebasto/bitcoin/compare/e6ebaa0a39fcbeea506d2cb20ad94b8b1476c61e..9e218cfca5b863ced0764da7af8b202965c0cd34) you pushed doesn't introduce `HAVE_BUILTIN_MUL_OVERFLOW`. Is this left for later?

hebasto replied:
> The [diff](https://github.com/hebasto/bitcoin/compare/e6ebaa0a39fcbeea506d2cb20ad94b8b1476c61e..9e218cfca5b863ced0764da7af8b202965c0cd34) you pushed doesn't introduce `HAVE_BUILTIN_MUL_OVERFLOW`. Is this left for later?

Yes, it is because it is a fuzz binary specific.

TheCharlatan replied:
Perfect, from my point of view all symbols are now accounted for :)

hebasto replied:
> For `HAVE_THREAD_LOCAL` see `target_compile_features( ... cxx_thread_local ... )`.

[Done](https://github.com/hebasto/bitcoin/pull/7#issuecomment-1439089362) in #7.

theuni commented on cmake/crc32c.cmake:
```diff
@@ -0,0 +1,117 @@
+
+target_include_directories(crc32c
+  PUBLIC
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/src/crc32c/include>
```
For other reviewers: This does not leak outside of leveldb because it includes this lib privately.

hebasto commented on cmake/crc32c.cmake:
```diff
@@ -0,0 +1,117 @@
+else()
+  set(SSE42_CXXFLAGS -msse4.2)
+endif()
+set(CMAKE_REQUIRED_FLAGS ${SSE42_CXXFLAGS})
```
Save an initial `CMAKE_REQUIRED_FLAGS`'s value before using of it, and restore the initial value after?

hebasto replied:
I mean, `cmake_reset_check_state()` ... `cmake_pop_check_state()` from [`CMakePushCheckState`](https://cmake.org/cmake/help/latest/module/CMakePushCheckState.html) module.

theuni replied:
Oooh yes +1, I'd much prefer that we start using this syntax from the beginning.

hebasto replied:
Implemented.

hebasto commented:
Addressed @TheCharlatan's [comment](https://github.com/hebasto/bitcoin/pull/6#discussion_r1113337263).

hebasto commented:
@theuni's [comment](https://github.com/hebasto/bitcoin/pull/6#discussion_r1113395531) has been addressed.

hebasto commented on cmake/crc32c.cmake:
```diff
@@ -0,0 +1,120 @@
+
+# Check for ARMv8 w/ CRC and CRYPTO extensions support in the compiler.
+set(ARM_CRC_CXXFLAGS -march=armv8-a+crc)
+set(CMAKE_REQUIRED_FLAGS ${ARM_CRC_CXXFLAGS})
```
Thanks! Reworked.

theuni commented:
> Addressed the recent @theuni's [comment](https://github.com/hebasto/bitcoin/pull/6#discussion_r1114698610).

Much better, thank you!

theuni commented:
Looking good to me. Will do a quick review of the whole thing tomorrow.

theuni commented on cmake/leveldb.cmake:
```diff
@@ -0,0 +1,94 @@
+# buildsystem.
+
+include(CheckCXXSymbolExists)
+check_cxx_symbol_exists(F_FULLFSYNC "fcntl.h" HAVE_FULLFSYNC)
```
We use this in `util/system.cpp` as well, makes sense to move it to introspection.

hebasto replied:
But it uses `defined(F_FULLFSYNC)` (see https://github.com/bitcoin/bitcoin/pull/3000), not `defined(HAVE_FULLFSYNC)`.

hebasto replied:
> But it uses `defined(F_FULLFSYNC)` (see [bitcoin#3000](https://github.com/bitcoin/bitcoin/pull/3000)), not `defined(HAVE_FULLFSYNC)`.

Looks like a bug, doesn't it?

hebasto replied:
FWIW, on my macOS systems, the `F_FULLFSYNC` is defined after `#include <fcntl.h>`.

theuni replied:
Yeah, this is kinda weird in our code so we can worry about it later.

vasild reviewed (approved):
ACK 79dfd773c561f8197ea0b57fcdafd0216e498134

I don't get it why this builds nothing (nor crc32c nor leveldb):

```sh
cmake -S . -B /tmp/b
make -C /tmp/b
# this does the compilation:
# make -C /tmp/b crc32c # or
# cmake --build /tmp/b --target crc32c
```

I guess I am missing something obvious, but my understanding is that [add_library()](https://cmake.org/cmake/help/latest/command/add_library.html) should add it to the default target.

hebasto commented:
@vasild 
> I don't get it why this builds nothing (nor crc32c nor leveldb):

Due to the [`EXCLUDE_FROM_ALL`](https://cmake.org/cmake/help/latest/prop_tgt/EXCLUDE_FROM_ALL.html#prop_tgt:EXCLUDE_FROM_ALL) option, the added targets are excluded fro the `all target.

That is why providing target explicitly is required.

hebasto commented:
Addressed @vasild's comments:
- https://github.com/hebasto/bitcoin/pull/6#discussion_r1115466851
- https://github.com/hebasto/bitcoin/pull/6#discussion_r1115468969

hebasto commented on cmake/introspection.cmake:
```diff
@@ -0,0 +1,281 @@
+  {
+    return posix_fallocate(0, 0, 0);
+  }
+  " HAVE_POSIX_FALLOCATE
```
Thanks! [Updated](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1441632392).

hebasto commented on cmake/introspection.cmake:
```diff
@@ -0,0 +1,281 @@
+  #include <sys/sysctl.h>
+
+  #ifdef __linux__
+  #error \"Don't use sysctl on Linux, it's deprecated even when it works\"
```
Thanks! [Updated](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1441632392).

hebasto commented on cmake/bitcoin-config.h.in:
```diff
@@ -30,6 +33,156 @@
+
+/* Define to 1 if you have the declaration of `be16toh', and to 0 if you
+   don't. */
+#cmakedefine01 HAVE_DECL_BE16TOH
```
In this particular case, CMake code is mimicking Autoconf's [`AC_CHECK_DECLS`](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.69/autoconf.html#index-AC_005fCHECK_005fDECLS-680) macro: https://github.com/hebasto/bitcoin/blob/17ddc033ba891a04777473c71fdf4b5d206049d0/configure.ac#L1014-L1019 

The possible lines in a `config/bitcoin-config.h` are:
- `#define HAVE_DECL_BE16TOH 1`
or
- `#define HAVE_DECL_BE16TOH 0`
which means that `HAVE_DECL_BE16TOH` is always defined.

CMake's [`#cmakedefine01`](https://cmake.org/cmake/help/latest/command/configure_file.html) does exactly the same.

hebasto replied:
**TL;DR: CMake's configure header**

1. `#define SYMBOL ...` -- `SYMBOL` is always defined.
2. `#cmakedefine01 SYMBOL` -- `SYMBOL` is always defined to a value of 0 or 1, depending on CMake's variable `SYMBOL`.
3. `#cmakedefine SYMBOL ...` --  `SYMBOL` can be defined or remain undefined, depending on CMake's variable `SYMBOL`.

vasild replied:
Yes, maybe at some point (out of the scope here) we should look into consistently only using `#if FOO` or only `#ifdef FOO` in the source code.

hebasto replied:
> Yes, maybe at some point (out of the scope here) we should look into consistently only using `#if FOO` or only `#ifdef FOO` in the source code.

Just FYI:
- https://github.com/bitcoin/bitcoin/issues/16419
- https://github.com/bitcoin/bitcoin/pull/25302

hebasto commented on cmake/crc32c.cmake:
```diff
@@ -0,0 +1,113 @@
+else()
+  set(SSE42_CXXFLAGS -msse4.2)
+endif()
+check_cxx_source_compiles_with_flags("${SSE42_CXXFLAGS}" "
+  #include <cstdint>
+  #if defined(_MSC_VER)
+  #include <intrin.h>
+  #elif defined(__GNUC__) && defined(__SSE4_2__)
+  #include <nmmintrin.h>
+  #endif
+
+  int main() {
+    uint64_t l = 0;
+    l = _mm_crc32_u8(l, 0);
+    l = _mm_crc32_u32(l, 0);
+    l = _mm_crc32_u64(l, 0);
+    return l;
+  }
+  " HAVE_SSE42
+)
```
The CMake's [`check_cxx_source_compiles()`](https://cmake.org/cmake/help/latest/module/CheckCXXSourceCompiles.html) command, along with other similar commands, uses a set of _global_ variables like `CMAKE_REQUIRED_FLAGS`, `CMAKE_REQUIRED_DEFINITIONS`, `CMAKE_REQUIRED_INCLUDES`, `CMAKE_REQUIRED_LINK_OPTIONS` and `CMAKE_REQUIRED_LIBRARIES`.

As @theuni pointed (https://github.com/hebasto/bitcoin/pull/6#discussion_r1113395531 and https://github.com/hebasto/bitcoin/pull/6#discussion_r1114698610), keeping changes of those variables local makes the build system less error prone.

Therefore, the `check_cxx_source_compiles_with_flags()` macro is a way to encapsulate changes of `CMAKE_REQUIRED_*` variables and keep the global environment unchanged.

vasild reviewed (approved):
ACK 17ddc033ba891a04777473c71fdf4b5d206049d0

Thanks for the explanations!

hebasto commented on cmake/introspection.cmake:
```diff
@@ -0,0 +1,281 @@
+
+check_include_file_cxx(string.h HAVE_STRING_H)
+if(HAVE_STRING_H)
+  check_cxx_source_compiles("
+    #include <string.h>
+
+    int main()
+    {
+      char buf[100];
+      char x = *strerror_r(0, buf, sizeof buf);
+      char* p = strerror_r(0, buf, sizeof buf);
+      return !p || x;
+    }
+    " STRERROR_R_CHAR_P
+  )
```
I took it from https://github.com/autotools-mirror/autoconf/blob/9d2637467df1d4811449dd14b2bebc142d9092b6/lib/autoconf/functions.m4#L1707-L1738.

Nevertheless, I'm going to consider and test your suggestion.

hebasto replied:
Hmm,
```c++
static_assert(std::is_same<decltype(strerror_r), char*(int, char*, size_t)>::value);
```
looks fragile as `noexcept` could be a part of the function type.

vasild replied:
Then it could be
```cpp
static_assert(std::is_same<decltype(strerror_r), char*(int, char*, size_t)>::value ||
              std::is_same<decltype(strerror_r), char*(int, char*, size_t) noexcept>::value);
```

hebasto replied:
> 1\. Why does the `return` value matter since `check_cxx_source_compiles()` is used?

To ["avoid unused variable warning"](https://github.com/autotools-mirror/autoconf/commit/3ddd7fba50272221ad09d9ae044b9891965fc122).

> 2\. Why call `strerror_r()` two times?

I guess that the first call is mostly for C language. Indeed, it is redundant in Bitcoin Core.

hebasto replied:
Thanks! [Updated](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1441845083).

hebasto replied:
> Also, I think both calls are equally C-ish or C++-ish.

I don't think that C compiler will fail if the return type of `func()` is a pointer to some other type.

TheCharlatan commented on cmake/introspection.cmake:
```diff
@@ -15,3 +17,243 @@ check_include_file_cxx(sys/prctl.h HAVE_SYS_PRCTL_H)
+)
+
+# - BSD getentropy()
+check_cxx_symbol_exists(getentropy "unistd.h;sys/random.h" HAVE_GETENTROPY_RAND)
```
We are sometimes using `check_cxx_source_compiles`, like with `SYS_getrandom` just above, and sometimes `check_cxx_symbols_exists` like here, while both `SYS_getrandom` and `getentropy` are checked with `AC_COMPILE_IFELSE` in autotools currently. What is logic that defines when either of the checks are used?

hebasto replied:
The [`check_cxx_symbol_exists`](https://cmake.org/cmake/help/latest/module/CheckCXXSymbolExists.html) command is intended to be used in simple cases.

Talking about the `SYS_getrandom`, it is called indirectly, and a more general `check_cxx_source_compiles()` command seems more appropriate here.

TheCharlatan replied:
Ok, I see that these criteria make sense here, e.g. we use `check_cxx_symbol_exists` for symbols with simple types called directly. Would it then make sense to use `check_cxx_symbol_exists` for `std::system` and `::_wsystem`?

hebasto replied:
> Would it then make sense to use `check_cxx_symbol_exists` for `std::system` and `::_wsystem`?

Thanks! [Updated](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1442023199).

hebasto commented:
Reworked check for `STRERROR_R_CHAR_P`. Thanks to @vasild :)

vasild reviewed (approved):
ACK 0a6cf6c42dc295ddef17eaf49a0c692486df1471

theuni reviewed (approved):
ACK 0a6cf6c42dc295ddef17eaf49a0c692486df1471

Thanks for being so receptive to changes here! I think it makes sense to be _really_ picky in the first PRs because precedent is being set for the rest. Hopefully review gets easier and easier :)

hebasto commented:
Addressed @TheCharlatan's [comment](https://github.com/hebasto/bitcoin/pull/6#discussion_r1115834383).

TheCharlatan reviewed (approved):
ACK df899456fbca934a3030572bb890aa00b3bcfe4c

> The resulted CMake's build/libleveldb.a is essentially the same as the combined Autotools' src/leveldb/.libs/libleveldb.a and src/leveldb/.libs/libmemenv.a.

The only user of memenv.h is dbwrapper.cpp, which also includes a bunch of the other leveldb includes. Nice change!

theuni commented:
> ACK [df89945](https://github.com/hebasto/bitcoin/commit/df899456fbca934a3030572bb890aa00b3bcfe4c)
> 
> > The resulted CMake's build/libleveldb.a is essentially the same as the combined Autotools' src/leveldb/.libs/libleveldb.a and src/leveldb/.libs/libmemenv.a.
> 
> The only user of memenv.h is dbwrapper.cpp, which also includes a bunch of the other leveldb includes. Nice change!

Note that [upstream has switched to building this way too](https://github.com/bitcoin/bitcoin/blob/master/src/leveldb/CMakeLists.txt#L219), so we're actually better off in that regard as well :)

ACK df899456fbca934a3030572bb890aa00b3bcfe4c.
